exonum/src/api/mod.rs

```diff
@@ -27,77 +28,65 @@ use iron::headers::Cookie;
 use hyper::header::{ContentType, SetCookie};
 use cookie::Cookie as CookiePair;
 use router::Router;
+use params;
 use serde_json;
 use serde::{Serialize, Serializer};
 use serde::de::{self, Visitor, Deserialize, Deserializer};
+use failure::Fail;
 
 use crypto::{PublicKey, SecretKey, Hash};
 use encoding::serialize::{FromHex, FromHexError, ToHex, encode_hex};
-use storage::{Result as StorageResult, Error as StorageError};
+use storage;
 
 pub mod public;
 pub mod private;
 #[cfg(test)]
 mod tests;
 
 /// List of possible Api errors.
-#[derive(Debug)]
+#[derive(Fail, Debug)]
 pub enum ApiError {
-    /// Service error.
-    Service(Box<::std::error::Error + Send + Sync>),
     /// Storage error.
-    Storage(StorageError),
-    /// Converting from hex error.
-    FromHex(FromHexError),
+    #[fail(display = "Storage error: {}", _0)]
+    Storage(
+        #[cause]
+        storage::Error
+    ),
+
     /// Input/output error.
-    Io(::std::io::Error),
+    #[fail(display = "IO error: {}", _0)]
+    Io(
+        #[cause]
+        ::std::io::Error
+    ),
+
     /// File not found.
+    #[fail(display = "File not found")]
     FileNotFound(Hash),
```

Comment:
May be off-topic, but `File*` errors have bugged me for a long time. It seems like they were introduced as a sort of a hack for some demo or project; IMO, they violate abstraction.

Reply:
Imo, this, the status code situation, and the copy-paste implementation of API (which I somewhat fixed here, but which is likely present in real-life servers as well) means that someone needs to take a holistic look at the HTTP API we expose and clean it up :-) I don't want to do it in this PR though, because it is orthogonal to "let's use failure and bury std::error::Error" goal :)

Reply:
This error is never used in code, why not to rename or remove it in this or following PR?

Reply:
>This error is never used in code, why not to rename or remove it in this or following PR?

Indeed 😆 ! I've just noticed that it is a dead code effectively! Will remove all unused options: if the consumer wants to have some custom IronError, they can just create a custion IronError directly :)